### ISSUE TITLE
There is a potential memory leaks in XSON when the detached objects are kept in RAM

### ISSUE THREAD

**Issue by miyconst** [Starcounter/Starcounter#4129]
Investigation is going on.

**Comment by chrhol** [comment]
Conclusions so far: the memory leak is not a memory leak per se, but one that happens due to a number of smaller things working badly together.

The main reason is the versioning for arrays that is used to allow client and server being out-of-sync with each other. This together with a pattern of having a client that rarely sends a request to the server, the server doing a lot of pushing, and that the session in some cases never times out makes the versionlog for old versions grow and grow.

**1. The versionlogs for arrays**
When versioning is enabled, each time the server bumps the version (for example doing a push) the current changes that have been made to an array is stored on the array. One set of changes for each version that is newer than the latest acknowledged clientversion. Old changesets are cleaned up when the client sends a patch, that confirms a version.

Currently, these changesets are stored including references to viewmodels and templates (this is due to the fact that the change-object is already created and just kept). This can be improved, since the only things needed is the operation that was done (add, remove, moved) and the index at that time. This will decrease the amount of memory needed to store the versionlogs.

**2. Cleaning up old logs**
Since the logs are only cleaned up when the client acknowledges a newer version (by sending a patch to the server), maybe we need to include the versioning in the keep-alive that is sent from the client. Today it's only makes sure that the session is touched. We still want to avoid a complete dirtycheck though.

**3. Session and timeout of session**
The session times out if not touched after a timeout (default 20 minutes). However the timestamp on the session is updated every time it is used, so if the server schedules a task for the session (`Session.ScheduleTask(...)`) this timestamp will be updated. 

Having a pattern with a timer that schedules a task for pushing data on a session, will make the session live for ever, even if the client has disconnected.

**Comment by andwah** [comment]
> 2. Cleaning up old logs

If i have an partial that displays a "clock" where the server updates the time by pushing out a new time each second. How can we clean up the logs in that case?. will a client "ping" be enough?

**Comment by chrhol** [comment]
> If i have an partial that displays a "clock" where the server updates the time by pushing out a new time each second. How can we clean up the logs in that case?. will a client "ping" be enough?

No, the ping will not be enough. The ping only sends an empty patch today (by design) to avoid triggering a complete dirtycheck. In this case you will have this problem that memory just keeps growing, until the client sends a proper patchrequest or memory runs out.

I think maybe we should enhance the ping to send the clients serverversion to allow cleaning up unneeded versioning info, but still not trigger a dirtycheck.

**Comment by chrhol** [comment]
@andwah: Just to clarify, if, every time you push you make changes in some array (i.e. add and remove items) then this would happen. If the only updates when you do the push is a timestamp for the clock then it will not be a problem.

**Comment by andwah** [comment]
@chrhol Thanks. Now i understand.

 In my case i make changes in arrays etc. and push them to a "passive" client.  so i'm looking forward for a solution.

**Comment by chrhol** [comment]
Issue for point 1: #4131 
Issue for point 2: #4130 

I think the third point is of a much lower prio, and maybe even works as it should. So lets wait with that one.
